synapse/util/caches/sync_response_cache.py

```diff
@@ -0,0 +1,101 @@
+import logging
+from typing import Any, Callable, Dict
+
+from twisted.internet import defer
+
+from synapse.util import Clock
+from synapse.util.async_helpers import ObservableDeferred
+from synapse.util.caches.response_cache import ResponseCache, T
+
+logger = logging.getLogger(__name__)
+
+
+# A special class for /sync responses, to conditionally cache these.
+class SyncResponseCache(ResponseCache[T]):
+    def __init__(self, clock: Clock, name: str, timeout_ms: float = 0):
+        super().__init__(clock, name, timeout_ms)
+
+        self.conditionals = {}  # type: Dict[T, Callable[[Any], bool]]
+
+    def run_conditional(self, key: T, result: Any) -> bool:
+        """Runs a conditional set on key T, defaults to True"""
+        cond = self.conditionals.get(key, None)
+        if cond is None:
+            return True
+        else:
+            try:
+                # Below type annotation is needed for mypy to shush about some statements being unreachable,
+                # we essentially have to not trust other functions to be able to correctly recover from any fallout
+                # (and log it)
+                res = cond(result)  # type: Any
+            except Exception:
+                logger.exception(
```

Comment:
`.exception` because then it throws it to sentry, i think this is pretty sentry-worthy